### PR TITLE
add JointPrimaryMarginalizedModel likelihood model

### PR DESCRIPTION
**Pull request by WuShichao**
@ahnitz This PR implements a likelihood model for multiband PE.

### CODE REVIEW

File: pycbc/inference/models/hierarchical.py [hunk 1]
```diff
@@ -599,3 +602,273 @@ def _loglikelihood(self):
             for stat in self.extra_stats_map[lbl]:
                 setattr(self._current_stats, stat, mstats[stat.subname])
         return logl
+
+
+class MultibandRelativeTimeDom(HierarchicalModel):
```
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:607]: It's not really time to the RelativeTimeDom, right? It's just tied to a primary model / secondary model scenario, no?

Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:607]: I named it this way because the primary model uses `RelativeTimeDom`, and also emphasizes the multiband purpose. I know you want a more general likelihood model though.

File: pycbc/inference/models/hierarchical.py [hunk 2]
```diff
@@ -599,3 +610,301 @@ def _loglikelihood(self):
...
+        hh_others = numpy.zeros(nums)
+
+        for _, other_model in enumerate(self.other_models):
+            current_params_other = other_model.current_params.copy()
+            print("current_params_other: ", current_params_other)
+            # TODO: run this for-loop in parallel
+            for i in range(nums):
+                current_params_other.update(
+                    {key: value[i] for key, value in margin_params.items()})
+                if 'distance' in self.primary_model.static_params:
```
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:710]: Don't hardcode specific things related to distance. Handle static parameters generally

Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:710]: See if you can make this more general

File: pycbc/inference/models/hierarchical.py [hunk 3]
```diff
@@ -599,3 +610,301 @@ def _loglikelihood(self):
...
+                # doesn't need marginalization
+                for p in wfparam_map[lbl]:
+                    submodel.static_params.pop(p.subname)
+            submodels[lbl] = submodel
+            logging.info("")
+
+        # remove all marginalized parameters from the top-level model's
+        # `variable_params` and `prior` sections
+        primary_model = '%s__model' % kwargs['primary_lbl'][0]
+        marginalized_params = cp.get(primary_model,
```
[the remark this answers is not in the document]
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py]: I have made this part more general. Please check.

File: pycbc/inference/models/hierarchical.py (hunk 2 above)
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py]: I have made this part more general. In order not to pass the primary model's `f_lower` to LISA, I need to add it into that `if` check.

File: pycbc/inference/models/hierarchical.py [hunk 4]
```diff
@@ -599,3 +603,298 @@ def _loglikelihood(self):
...
+        sampattrs = fp.getattrs(group=fp.samples_group)
+        # if a group is specified, prepend the lognl names with it
+        if group is None or group == '/':
+            prefix = ''
+        else:
+            prefix = group.replace('/', '__')
+            if not prefix.endswith('__'):
+                prefix += '__'
+        try:
+            for lbl, model in self.submodels.items():
```
Comment (WuShichao) [inline pycbc/inference/models/hierarchical.py:656]: @ahnitz I also modified this `write_metadata` method, previously I just copied it from that `MultiSignalModel`, but I noticed there is a main difference between my model and it: doesn't like `MultiSignalModel`, my submodels don't share the same detector data, so I have to save all `lognl`. Is this part correct?

File: pycbc/inference/models/hierarchical.py [hunk 5]
```diff
@@ -599,3 +603,356 @@ def _loglikelihood(self):
...
+                sampattrs['{}lognl'.format(prefix + '%s__' % lbl)
```
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:657]: You should be able to avoid duplicating this code.

File: pycbc/inference/models/hierarchical.py (hunk 1 above)
Comment (ahnitz) [reply pycbc/inference/models/hierarchical.py:608]: @WuShichao I don't think this is tied to the relative model, so I think the name still needs to be fixed. (it is tied to models that do marginalization).

Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:608]: @ahnitz Can you suggest a name? This multiband likelihood relies on the primary model's `RelativeTimeDom`, and has specific code to do the marginalization for LISA. So it's not a likelihood model for general usage, right?

Comment (ahnitz) [reply pycbc/inference/models/hierarchical.py:608]: It doesn't is just relies on it doing vector maginalization. Nothing explicitly requires that either be a LISA model or even be specific models, just that the primary use vector marginalization at some point. 

It could be used for general useage if for example you want to do joint EM + GW analysis. In that case, sky marginalization would still be useful and one could just use your code with modified configuration files.

Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:608]: @ahnitz OK, then let's name it `JointPrimaryModelMarginalized`, do you think it's better?

File: pycbc/inference/models/hierarchical.py [hunk 6]
```diff
@@ -113,17 +115,18 @@ def __init__(self, variable_params, submodels, **kwargs):
...
-                                 "in a hierarchical analysis, these are "
-                                 "handled by the hiearchical model"
-                                 .format(lbl))
-            if model.waveform_transforms is not None:
-                raise ValueError("Model {} has waveform transforms set; "
-                                 "in a hierarchical analysis, these are "
-                                 "handled by the hiearchical model"
-                                 .format(lbl))
+            # model, except for `joint_primary_marginalized` model
+            if self.name != "joint_primary_marginalized":
```
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:119]: Why is this here? Ideally anything different only exists in each subclass. Parent classes shouldn't have know about subclasses really.

File: pycbc/inference/models/tools.py [hunk 7]
```diff
@@ -766,7 +770,12 @@ def get_loglr():
             self.reconstruct_phase = False
 
         rec['loglr'] = loglr[xl]
-        rec['loglikelihood'] = self.lognl + rec['loglr']
+        if set_others_lognl is None:
```
Comment (ahnitz) [inline pycbc/inference/models/tools.py:773]: This could probably be handled in your model since lognl doesn't affect the drawing of anything (it's a constant in your model), no?

Comment (ahnitz) [reply pycbc/inference/models/tools.py:773]: That would mean you wouldn not need to pass this in, just update it in your method.

File: pycbc/inference/models/hierarchical.py (hunk 6 above)
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:121]: @ahnitz In the standard `hierarchical` model, all the transforms will be handled on the top level, but in our case, remember that we need to allow the transform with the prefix to be handled by submodels. Currently, I don't have a better idea. We can fix this in later patches.

Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:121]: @ahnitz In the standard `hierarchical` model, all the transforms will be handled on the top level, but in our case, remember that we need to allow the transform with the prefix to be handled by submodels. Currently, I don't have a better idea. We can fix this in later patches.

Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:121]: @ahnitz In the standard `hierarchical` model, all the transforms will be handled on the top level, but in our case, remember that we need to allow the transform with the prefix to be handled by submodels. Currently, I don't have a better idea. We can fix this in later patches.

File: pycbc/inference/models/hierarchical.py [hunk 8]
```diff
@@ -599,3 +604,316 @@ def _loglikelihood(self):
...
+        sh_others = numpy.full(nums, 0 + 0.0j)
+        hh_others = numpy.zeros(nums)
+
+        for _, other_model in enumerate(self.other_models):
+            current_params_other = other_model.current_params.copy()
+            for i in range(nums):
+                current_params_other.update(
+                    {key: value[i] for key, value in margin_params.items()})
+                for p in self.primary_model.static_params.keys():
+                    # don't pass primary model's 'f_lower' to LISA
```
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:707]: Why are you copying any static params to LISA? Shouldn't the normal hierarchical handling already handle static params?

File: pycbc/inference/models/hierarchical.py [hunk 9]
```diff
@@ -599,3 +604,316 @@ def _loglikelihood(self):
...
+        logging.info("Loading joint_primary_marginalized model")
+        return super(HierarchicalModel, cls).from_config(
+                cp, submodels=submodels, **kwargs)
+
+    def reconstruct(self, seed=None):
+        """ Reconstruct marginalized parameters by using the primary
+        model's reconstruct method, total_loglr and others_lognl.
+        """
+        rec = self.primary_model.reconstruct(
+                seed=seed, set_loglr=self.total_loglr())
```
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:916]: How is this supposed to work? You are just calling self.total_loglr and then passing in the return? Don't you need to pass in a *function* that can be called in the reconstruct method?

File: pycbc/inference/models/hierarchical.py (hunk 8 above)
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:707]: @ahnitz If I remember correctly, I did this because I wanted to avoid some of 3G's static parameters, such `f_lower` to be used by LISA, so just overwrite it here to make sure.

File: pycbc/inference/models/tools.py [hunk 10]
```diff
@@ -728,11 +728,14 @@ def reconstruct(self, rec=None, seed=None):
         if rec is None:
             rec = {}
 
-        def get_loglr():
-            p = self.current_params.copy()
-            p.update(rec)
-            self.update(**p)
-            return self.loglr
+        if set_loglr is None:
```
Comment (ahnitz) [inline pycbc/inference/models/tools.py:731]: @WuShichao The flags used here interact with the output of the loglr function. You should make sure you undersand how that works to determine how the function you pass in should behave. Will it behave correctly in these cases? You know the primary model on its own will, but you seem to pass your models loglr function. Does it respect these flags properly? I'm not sure your solution has through through this yet.

File: pycbc/inference/models/hierarchical.py (hunk 9 above)
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py]: Now I almost exactly follow the structure of the original `recntruct`, define a `get_loglr` function to call my `total_loglr()`, my `self.total_loglr()` corresponding to `self.loglr`.

File: pycbc/inference/models/tools.py (hunk 10 above)
Comment (WuShichao) [reply pycbc/inference/models/tools.py:731]: Now I pass a function to it.

File: pycbc/inference/models/hierarchical.py (hunk 9 above)
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py]: I forgot the difference between `set_loglr=get_loglr` and `set_loglr=get_loglr()`, now I know we need to use  `set_loglr=get_loglr` if we want to pass a _function_ to it.

File: pycbc/inference/models/hierarchical.py [hunk 11]
```diff
@@ -599,3 +604,349 @@ def _loglikelihood(self):
...
+
+        # note that for SOBHB signals, ground-based detectors dominant SNR
+        # and accuracy of (tc, ra, dec)
+        self.primary_model.return_sh_hh = True
+        sh_primary, hh_primary = self.primary_model.loglr
+        self.primary_model.return_sh_hh = False
+
+        margin_names = list(
+            self.primary_model.marginalize_vector_params.keys())
+        margin_names.remove('logw_partial')
```
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:695]: Do you need this removal? What happens if you don't do this?

**Review by ahnitz — APPROVED** [review]
@WuShichao If you are now confident this works in your tests, you may merge. Please see the few remaining comments to see if they can be easily addressed, otherwise go ahead.

File: pycbc/inference/models/hierarchical.py (hunk 11 above)
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py]: @ahnitz Those are related to updating parameters in the LISA model, so there is no need for `logw_partial`  in the LISA model.

File: pycbc/inference/models/hierarchical.py (hunk 4 above)
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:663]: @ahnitz Just checked, the numbers saved in the final posterior file makes sense:
image
image

**Review by ahnitz — APPROVED** [review]
@WuShichao Add a comment as we discussed and then merge once the tests pass
